### PR TITLE
Extend CBMC Viewer to work with RBMC

### PR DESCRIPTION
**Pull request by markrtuttle**
This pull request make three small changes required for viewer to work on the output of rbmc in addition to cbmc.

### CODE REVIEW

File: compare-coverage.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        dict2 = coverage2['viewer-coverage']
+
+        if full:
+            return dict1 == dict2
+
+        return dict1["line_coverage"] == dict2["line_coverage"]
+    except KeyError:
+        return False
+
+################################################################
```
Comment (feliperodri) [inline compare-coverage.py:100]: I'd remove it, seems unnecessary,

File: compare-coverage.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    return parser
+
+def dict_diff(dict1, dict2):
+        with open('/tmp/coverage1.json', 'w') as out1:
+            print(json.dumps(dict1, indent=2, sort_keys=True), file=out1)
+        with open('/tmp/coverage2.json', 'w') as out2:
+            print(json.dumps(dict2, indent=2, sort_keys=True), file=out2)
+
+        proc = subprocess.run(
+            ['diff', '/tmp/coverage1.json', '/tmp/coverage2.json'],
```
Comment (feliperodri) [inline compare-coverage.py:36]: What about [filecmp](https://docs.python.org/3/library/filecmp.html)?

File: compare-coverage.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+                print(json.dumps(idict2, indent=2, sort_keys=True), file=out2)
+
+            proc = subprocess.run(
+                ['diff', '/tmp/coverage1.json', '/tmp/coverage2.json'],
+                text=True,
+                capture_output=True,
+                check=False
+            )
+            if proc.stdout:
+                difference[key] = proc.stdout
```
Comment (feliperodri) [inline compare-coverage.py:75]: What don't you call `dict_diff` here?

File: compare-result.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        keys = set([*dict1.keys(), *dict2.keys()])
+        if not full:
+            keys.remove('program')
+            keys.remove('status')
+
+        return all([dict1[key] == dict2[key] for key in keys])
+    except KeyError:
+        return False
+
+################################################################
```
Comment (feliperodri) [inline compare-result.py:43]: I'd remove it. It seems unnecessary.

**Review by feliperodri — APPROVED** [review]
LGTM module a few comments/questions.

File: cbmc_viewer/markup_summary.py [hunk 5]
```diff
@@ -134,7 +134,7 @@ def function_coverage(coverage, symbols):
             'total': func_cov['total'],
             'file_name': file_name,
             'func_name': func_name,
-            'line_num': symbols.lookup(func_name)["line"]
+            'line_num': symbols.lookup(func_name).get("line") if symbols.lookup(func_name) else 0
```
Comment (danielsn) [inline cbmc_viewer/markup_summary.py:137]: are we guaranteed that `line` exists if `func_name` does?

File: compare-coverage.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    parser.add_argument(
+        'coverage2',
+        help="""
+        JSON coverage output of make-coverage or cbmc-viewer.
+        """
+    )
+    return parser
+
+def dict_diff(dict1, dict2):
+        with open('/tmp/coverage1.json', 'w') as out1:
```
Comment (danielsn) [inline compare-coverage.py:30]: Should we ask python for a true tempfile? Hardcoding the names will cause problems if we run more than one process at once.

File: compare-coverage.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    )
+    parser.add_argument(
+        'coverage2',
+        help="""
+        JSON coverage output of make-coverage or cbmc-viewer.
+        """
+    )
+    return parser
+
+def dict_diff(dict1, dict2):
```
Comment (danielsn) [inline compare-coverage.py:29]: It was a little unclear what this function does - took me several reads to understand. A comment would be helpful.

File: compare-coverage.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+            idict1 = {int(line): value for line, value in dict1[key].items()}
+            with open('/tmp/coverage1.json', 'w') as out1:
+                print(json.dumps(idict1, indent=2, sort_keys=True), file=out1)
+
+            idict2 = {int(line): value for line, value in dict2[key].items()}
+            with open('/tmp/coverage2.json', 'w') as out2:
+                print(json.dumps(idict2, indent=2, sort_keys=True), file=out2)
+
+            proc = subprocess.run(
+                ['diff', '/tmp/coverage1.json', '/tmp/coverage2.json'],
```
Comment (danielsn) [inline compare-coverage.py:69]: are we guaranteed this will be on all machines we want to run on?

File: compare-coverage.py [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+            if proc.stdout:
+                difference[key] = proc.stdout
+
+        if not dict1_missing and not dict2_missing and not difference:
+            return {}
+
+        return {'left-missing': sorted(dict1_missing),
+                'right-missing': sorted(dict2_missing),
+                'difference': {key: value
+                               for key, value in sorted(difference.items())}
```
Comment (danielsn) [inline compare-coverage.py:83]: why not just make the map up front?

File: compare-coverage.py [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+                }
+    except KeyError:
+        raise UserWarning("Diff computation failed.")
+
+def compare(coverage1, coverage2, full=False, verbose=True):
```
Comment (danielsn) [inline compare-coverage.py:88]: Comment explaining what `full` means?

Comment (danielsn) [inline compare-coverage.py:88]: I believe verbose is not used?

File: compare-result.py [hunk 11]
```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env python3
```
Comment (danielsn) [inline compare-result.py:1]: Are we really not putting copyright headers on things?

File: compare-result.py [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    """Command line parser."""
+
+    parser = argparse.ArgumentParser(
+        description='Compare CBMC property checking results.'
+    )
+
+    parser.add_argument(
+        'result1',
+        help="""
+        JSON result output of make-result or cbmc-viewer.
```
Comment (danielsn) [inline compare-result.py:17]: what is `make-result`?

File: compare-result.py [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+        dict1 = result1['viewer-result']
+        dict2 = result2['viewer-result']
+
+        keys = set([*dict1.keys(), *dict2.keys()])
+        if not full:
+            keys.remove('program')
+            keys.remove('status')
+
+        return all([dict1[key] == dict2[key] for key in keys])
+    except KeyError:
```
Comment (danielsn) [inline compare-result.py:40]: Maybe this is pythonic, in which case keep doing it. But using exceptions for ordinary control flow seems weird to my non python native brain.

**Review by danielsn — CHANGES_REQUESTED** [review]
Nothing major. But a few suggestions.

File: compare-coverage.py (hunk 1 above)
Comment (markrtuttle) [reply compare-coverage.py:100]: Done.

File: compare-result.py (hunk 4 above)
Comment (markrtuttle) [reply compare-result.py:43]: Done.

**Comment by markrtuttle** [comment]
Closed in favor of #33

**Comment by markrtuttle** [comment]
The comparison scripts included in this pull request that generated the bulk of the discussion will be addressed in a separate pull request.

**Pull request closed without merging**